File: backend/internal/service/tariff_service.go
package service

import (
	"errors"
	"fmt"
	"time"

	"github.com/fastenmind/fastener-api/internal/models"
	"github.com/fastenmind/fastener-api/internal/repository"
	"github.com/google/uuid"
)

type TariffService interface {
	// HS Codes
	SearchHSCodes(params map[string]interface{}) ([]models.HSCode, int64, error)
	GetHSCode(code string) (*models.HSCode, error)
	ValidateHSCode(code, country string) (bool, string, []models.HSCode, error)
	GetCommonHSCodes(category string) ([]models.HSCode, error)
	
	// Tariff Calculation
	CalculateTariff(req TariffCalculationRequest) (*TariffCalculationResult, error)
	BatchCalculateTariff(items []TariffCalculationRequest) ([]TariffCalculationResult, error)
	
	// Trade Agreements
	GetTradeAgreements(countries []string) ([]models.TradeAgreement, error)
	
	// History
	GetCalculationHistory(companyID uuid.UUID, limit int) ([]models.TariffCalculation, error)
}

type TariffCalculationRequest struct {
	CompanyID              uuid.UUID `json:"-"`
	UserID                 uuid.UUID `json:"-"`
	HSCode                 string    `json:"hs_code"`
	FromCountry            string    `json:"from_country"`
	ToCountry              string    `json:"to_country"`
	ProductValue           float64   `json:"product_value"`
	Quantity               float64   `json:"quantity"`
	Unit                   string    `json:"unit"`
	WeightKG               float64   `json:"weight_kg,omitempty"`
	Currency               string    `json:"currency"`
	Incoterm               string    `json:"incoterm,omitempty"`
	PreferentialTreatment  bool      `json:"preferential_treatment"`
}

type TariffCalculationResult struct {
	HSCode             string                 `json:"hs_code"`
	FromCountry        string                 `json:"from_country"`
	ToCountry          string                 `json:"to_country"`
	ProductValue       float64                `json:"product_value"`
	TariffRate         *models.TariffRate     `json:"tariff_rate"`
	CalculatedTariff   float64                `json:"calculated_tariff"`
	EffectiveRate      float64                `json:"effective_rate"`
	Currency           string                 `json:"currency"`
	CalculationDetails CalculationDetails     `json:"calculation_details"`
	Warnings           []string               `json:"warnings,omitempty"`
}

type CalculationDetails struct {
	BaseValue             float64 `json:"base_value"`
	AdValoremDuty         float64 `json:"ad_valorem_duty,omitempty"`
	SpecificDuty          float64 `json:"specific_duty,omitempty"`
	TotalDuty             float64 `json:"total_duty"`
	PreferentialApplied   bool    `json:"preferential_applied"`
	PreferentialSavings   float64 `json:"preferential_savings,omitempty"`
}

type tariffService struct {
	repo repository.TariffRepository
}

func NewTariffService(repo repository.TariffRepository) TariffService {
	return &tariffService{repo: repo}
}

func (s *tariffService) SearchHSCodes(params map[string]interface{}) ([]models.HSCode, int64, error) {
	return s.repo.FindHSCodes(params)
}

func (s *tariffService) GetHSCode(code string) (*models.HSCode, error) {
	return s.repo.GetHSCode(code)
}

func (s *tariffService) ValidateHSCode(code, country string) (bool, string, []models.HSCode, error) {
	// Check if HS code exists
	hsCode, err := s.repo.GetHSCode(code)
	if err != nil {
		// If not found, suggest similar codes
		suggestions, _, _ := s.repo.FindHSCodes(map[string]interface{}{
			"search":    code[:4], // Search by first 4 digits
			"is_active": true,
			"page_size": 5,
		})
		return false, "HS Code not found", suggestions, nil
	}
	
	if !hsCode.IsActive {
		return false, "HS Code is inactive", nil, nil
	}
	
	// Country-specific validation could be added here
	// For now, just return valid
	return true, "", nil, nil
}

func (s *tariffService) GetCommonHSCodes(category string) ([]models.HSCode, error) {
	params := map[string]interface{}{
		"is_active": true,
		"page_size": 20,
	}
	
	if category != "" {
		params["category"] = category
	}
	
	codes, _, err := s.repo.FindHSCodes(params)
	return codes, err
}

func (s *tariffService) CalculateTariff(req TariffCalculationRequest) (*TariffCalculationResult, error) {
	// Validate input
	if req.HSCode == "" || req.FromCountry == "" || req.ToCountry == "" {
		return nil, errors.New("HS code, from country, and to country are required")
	}
	
	if req.ProductValue <= 0 {
		return nil, errors.New("product value must be greater than 0")
	}
	
	// Get effective tariff rate
	now := time.Now()
	rate, err := s.repo.GetEffectiveTariffRate(req.HSCode, req.FromCountry, req.ToCountry, now)
	if err != nil {
		// If no specific rate found, try to find general rate or return zero tariff
		return s.calculateWithZeroTariff(req, "No tariff rate found for this route")
	}
	
	// Calculate tariff based on rate type
	result := &TariffCalculationResult{
		HSCode:       req.HSCode,
		FromCountry:  req.FromCountry,
		ToCountry:    req.ToCountry,
		ProductValue: req.ProductValue,
		TariffRate:   rate,
		Currency:     req.Currency,
		Warnings:     []string{},
	}
	
	details := CalculationDetails{
		BaseValue: req.ProductValue,
	}
	
	// Check for preferential treatment
	effectiveRate := rate.RateValue
	if req.PreferentialTreatment && rate.PreferentialRate > 0 {
		originalRate := effectiveRate
		effectiveRate = rate.PreferentialRate
		details.PreferentialApplied = true
		
		// Add conditions warning if any
		if rate.PreferentialConditions != "" {
			result.Warnings = append(result.Warnings, 
				fmt.Sprintf("Preferential rate conditions: %s", rate.PreferentialConditions))
		}
	}
	
	// Calculate based on rate type
	switch rate.RateType {
	case "ad_valorem":
		// Percentage of value
		details.AdValoremDuty = req.ProductValue * (effectiveRate / 100)
		details.TotalDuty = details.AdValoremDuty
		
	case "specific":
		// Fixed amount per unit
		if req.Quantity <= 0 {
			return nil, errors.New("quantity required for specific duty calculation")
		}
		details.SpecificDuty = rate.SpecificRate * req.Quantity
		details.TotalDuty = details.SpecificDuty
		
	case "compound":
		// Both ad valorem and specific
		details.AdValoremDuty = req.ProductValue * (effectiveRate / 100)
		if req.Quantity > 0 && rate.SpecificRate > 0 {
			details.SpecificDuty = rate.SpecificRate * req.Quantity
		}
		details.TotalDuty = details.AdValoremDuty + details.SpecificDuty
		
	default:
		return nil, fmt.Errorf("unknown rate type: %s", rate.RateType)
	}
	
	// Calculate preferential savings if applied
	if details.PreferentialApplied && rate.RateValue != rate.PreferentialRate {
		normalDuty := req.ProductValue * (rate.RateValue / 100)
		details.PreferentialSavings = normalDuty - details.TotalDuty
		if details.PreferentialSavings < 0 {
			details.PreferentialSavings = 0
		}
	}
	
	result.CalculatedTariff = details.TotalDuty
	result.EffectiveRate = details.TotalDuty / req.ProductValue
	result.CalculationDetails = details
	
	// Add warnings based on incoterm
	if req.Incoterm != "" {
		switch req.Incoterm {
		case "DDP":
			result.Warnings = append(result.Warnings, "DDP: Seller bears all tariff costs")
		case "DDU", "DAP":
			result.Warnings = append(result.Warnings, "Buyer responsible for import duties")
		}
	}
	
	// Save calculation record
	calcRecord := &models.TariffCalculation{
		CompanyID:             req.CompanyID,
		UserID:                req.UserID,
		HSCode:                req.HSCode,
		FromCountry:           req.FromCountry,
		ToCountry:             req.ToCountry,
		ProductValue:          req.ProductValue,
		Quantity:              req.Quantity,
		Unit:                  req.Unit,
		WeightKG:              req.WeightKG,
		Currency:              req.Currency,
		Incoterm:              req.Incoterm,
		PreferentialTreatment: req.PreferentialTreatment,
		TariffRateID:          &rate.ID,
		CalculatedTariff:      result.CalculatedTariff,
		EffectiveRate:         result.EffectiveRate,
		CalculationDetails:    details,
		Warnings:              result.Warnings,
	}
	
	if err := s.repo.CreateCalculation(calcRecord); err != nil {
		// Log error but don't fail the calculation
		result.Warnings = append(result.Warnings, "Failed to save calculation history")
	}
	
	return result, nil
}

func (s *tariffService) calculateWithZeroTariff(req TariffCalculationRequest, warning string) (*TariffCalculationResult, error) {
	result := &TariffCalculationResult{
		HSCode:           req.HSCode,
		FromCountry:      req.FromCountry,
		ToCountry:        req.ToCountry,
		ProductValue:     req.ProductValue,
		CalculatedTariff: 0,
		EffectiveRate:    0,
		Currency:         req.Currency,
		CalculationDetails: CalculationDetails{
			BaseValue: req.ProductValue,
			TotalDuty: 0,
		},
		Warnings: []string{warning},
	}
	
	return result, nil
}

func (s *tariffService) BatchCalculateTariff(items []TariffCalculationRequest) ([]TariffCalculationResult, error) {
	results := make([]TariffCalculationResult, 0, len(items))
	
	for _, item := range items {
		result, err := s.CalculateTariff(item)
		if err != nil {
			// Add error as warning and continue
			result = &TariffCalculationResult{
				HSCode:       item.HSCode,
				FromCountry:  item.FromCountry,
				ToCountry:    item.ToCountry,
				ProductValue: item.ProductValue,
				Currency:     item.Currency,
				Warnings:     []string{fmt.Sprintf("Calculation error: %v", err)},
			}
		}
		results = append(results, *result)
	}
	
	return results, nil
}

func (s *tariffService) GetTradeAgreements(countries []string) ([]models.TradeAgreement, error) {
	return s.repo.FindTradeAgreements(countries)
}

func (s *tariffService) GetCalculationHistory(companyID uuid.UUID, limit int) ([]models.TariffCalculation, error) {
	return s.repo.GetCalculationHistory(companyID, limit)
}

File: frontend/src/services/tariff.service.ts
import api from './api'

export interface HSCode {
  id: string
  code: string
  description: string
  description_en?: string
  unit: string
  category: string
  parent_code?: string
  is_active: boolean
}

export interface TariffRate {
  id: string
  hs_code: string
  from_country: string
  to_country: string
  rate_type: 'ad_valorem' | 'specific' | 'compound'
  rate_value: number
  specific_rate?: number
  currency?: string
  unit?: string
  effective_from: string
  effective_to?: string
  preferential_rate?: number
  preferential_conditions?: string
  notes?: string
}

export interface TariffCalculationRequest {
  hs_code: string
  from_country: string
  to_country: string
  product_value: number
  quantity: number
  unit: string
  weight_kg?: number
  currency: string
  incoterm?: string
  preferential_treatment?: boolean
}

export interface TariffCalculationResult {
  hs_code: string
  from_country: string
  to_country: string
  product_value: number
  tariff_rate: TariffRate
  calculated_tariff: number
  effective_rate: number
  currency: string
  calculation_details: {
    base_value: number
    ad_valorem_duty?: number
    specific_duty?: number
    total_duty: number
    preferential_applied: boolean
    preferential_savings?: number
  }
  warnings?: string[]
}

export interface CountryTradeAgreement {
  id: string
  agreement_name: string
  member_countries: string[]
  effective_date: string
  preferential_rates: boolean
  certificate_required: boolean
  description?: string
}

class TariffService {
  async searchHSCodes(params: {
    search?: string
    category?: string
    page?: number
    page_size?: number
  }): Promise<{ data: HSCode[]; pagination: any }> {
    const response = await api.get('/tariffs/hs-codes', { params })
    return response.data
  }

  async getHSCode(code: string): Promise<HSCode> {
    const response = await api.get(`/tariffs/hs-codes/${code}`)
    return response.data
  }

  async getTariffRates(params: {
    hs_code: string
    from_country?: string
    to_country?: string
  }): Promise<TariffRate[]> {
    const response = await api.get('/tariffs/rates', { params })
    return response.data
  }

  async calculateTariff(data: TariffCalculationRequest): Promise<TariffCalculationResult> {
    const response = await api.post('/tariffs/calculate', data)
    return response.data
  }

  async batchCalculate(
    items: TariffCalculationRequest[]
  ): Promise<TariffCalculationResult[]> {
    const response = await api.post('/tariffs/batch-calculate', { items })
    return response.data
  }

  async getTradeAgreements(countries: string[]): Promise<CountryTradeAgreement[]> {
    const response = await api.get('/tariffs/trade-agreements', {
      params: { countries: countries.join(',') }
    })
    return response.data
  }

  async validateHSCode(code: string, country: string): Promise<{
    valid: boolean
    message?: string
    suggested_codes?: HSCode[]
  }> {
    const response = await api.post('/tariffs/validate-hs-code', { code, country })
    return response.data
  }

  async getCommonHSCodes(category?: string): Promise<HSCode[]> {
    const response = await api.get('/tariffs/common-hs-codes', { params: { category } })
    return response.data
  }

  // Static data for demo/testing
  async getCountryList(): Promise<Array<{ code: string; name: string; region: string }>> {
    return [
      { code: 'TW', name: '台灣 Taiwan', region: 'Asia' },
      { code: 'CN', name: '中國 China', region: 'Asia' },
      { code: 'US', name: '美國 United States', region: 'Americas' },
      { code: 'DE', name: '德國 Germany', region: 'Europe' },
      { code: 'JP', name: '日本 Japan', region: 'Asia' },
      { code: 'KR', name: '韓國 South Korea', region: 'Asia' },
      { code: 'VN', name: '越南 Vietnam', region: 'Asia' },
      { code: 'TH', name: '泰國 Thailand', region: 'Asia' },
      { code: 'MY', name: '馬來西亞 Malaysia', region: 'Asia' },
      { code: 'SG', name: '新加坡 Singapore', region: 'Asia' },
      { code: 'GB', name: '英國 United Kingdom', region: 'Europe' },
      { code: 'FR', name: '法國 France', region: 'Europe' },
      { code: 'IT', name: '義大利 Italy', region: 'Europe' },
      { code: 'MX', name: '墨西哥 Mexico', region: 'Americas' },
      { code: 'CA', name: '加拿大 Canada', region: 'Americas' },
      { code: 'AU', name: '澳洲 Australia', region: 'Oceania' },
      { code: 'IN', name: '印度 India', region: 'Asia' },
      { code: 'BR', name: '巴西 Brazil', region: 'Americas' },
    ]
  }

  // Common HS codes for fasteners
  async getFastenerHSCodes(): Promise<Array<{ code: string; description: string }>> {
    return [
      { code: '7318.11', description: '鐵或鋼製螺栓 (Coach screws)' },
      { code: '7318.12', description: '鐵或鋼製其他木螺釘 (Other wood screws)' },
      { code: '7318.13', description: '鐵或鋼製鉤頭螺釘及環頭螺釘 (Screw hooks and screw rings)' },
      { code: '7318.14', description: '鐵或鋼製自攻螺釘 (Self-tapping screws)' },
      { code: '7318.15', description: '鐵或鋼製其他螺釘及螺栓 (Other screws and bolts)' },
      { code: '7318.16', description: '鐵或鋼製螺帽 (Nuts)' },
      { code: '7318.19', description: '鐵或鋼製其他螺紋製品 (Other threaded articles)' },
      { code: '7318.21', description: '鐵或鋼製彈簧墊圈及其他防鬆墊圈 (Spring washers and other lock washers)' },
      { code: '7318.22', description: '鐵或鋼製其他墊圈 (Other washers)' },
      { code: '7318.23', description: '鐵或鋼製鉚釘 (Rivets)' },
      { code: '7318.24', description: '鐵或鋼製開口銷及開尾銷 (Cotters and cotter-pins)' },
      { code: '7318.29', description: '鐵或鋼製其他無螺紋製品 (Other non-threaded articles)' },
    ]
  }
}

export const tariffService = new TariffService()
export default tariffService